### DISCUSSION TITLE
Runtime kernel compiling

### DISCUSSION THREAD

**Discussion by trsh** [LuisaGroup/LuisaCompute#151] [Q&A]
I wonder if this is possible. Given a kernel as string, compiling it at runtime and then executing.

**Comment by Mike-Leo-Smith** [comment]
It's possible since all backends generate the shader code first and then compile it at runtime. But currently we don't expose such APIs. There might be some little glitches to support such a feature: the backends expect specific kernel argument layouts (or shader descriptor sets, etc.), to which the user provided kernel strings must also conform.

An alternative way (which I personally would prefer) might be directly using the target graphics API (like native CUDA/DirectX) and interoperate with the LuisaCompute system with its exposed native handles. For example, you can get the underlying CUDA buffer address by reinterpreting Luisa's `Buffer<T>::native_handle()` to `CUdeviceptr` and use it with any native CUDA runtime/driver API such as `cuMemcpyAsync`.

**Reply by trsh** [reply]
I am not sure how your advice would help me. Maybe I should describe the use case: User inputs kernel code as text in some app, prior knowing available bindings of course (those are pretty much hardcoded), this gets compiled and runs. Think of Shadertoy, but not on web and with c++ kernel code instead of glsl shader.

**Reply by Mike-Leo-Smith** [reply]
Hi @trsh 

Thanks for the additional info. If I understand it correctly, it's something like hot-loading a shader (or a part of a shader)? And when users change the code, the shader gets recompiled and reflects the changes?

**Reply by trsh** [reply]
Yeah
